src/arbcall.jl

```diff
@@ -163,76 +168,66 @@ function jlfname(
-            )
-                :(precision($(Symbol(name(a)))))
-            else
-                :(DEFAULT_PRECISION[])
-            end
-        push!(kwargs, Expr(:kw, :($p::Integer), default))
-        deleteat!(arg_names, k)
-        deleteat!(c_types, k)
-        deleteat!(jl_types, k)
-    end
-
-    k = findfirst(==(:len), arg_names)
-    if !isnothing(k)
-        @assert c_types[k] == Clong
-        len = :len
-        a = first(cargs)
-        if jltype(a) ∈ (
-            Union{ArbVector,cstructtype(ArbVector),Ptr{arb_struct}},
-            Union{AcbVector,cstructtype(AcbVector),Ptr{acb_struct}},
-        )
-            push!(kwargs, Expr(:kw, :($len::Integer), :(length($(Symbol(name(a)))))))
-
-            deleteat!(arg_names, k)
-            deleteat!(c_types, k)
-            deleteat!(jl_types, k)
+    prec_kwarg = false
+    rnd_kwarg = false
+    len_keywords = Set{Symbol}()
+    for (i, carg) in enumerate(cargs)
+        if !argument_detection
+            push!(jl_arg_names_types, (name(carg), jltype(carg)))
+            continue
         end
-    end
 
-    k = findfirst(==(:rnd), arg_names)
-    if !isnothing(k)
-        @assert c_types[k] == arb_rnd || c_types[k] == Base.MPFR.MPFRRoundingMode
-        r = :rnd
-        if c_types[k] == arb_rnd
-            push!(kwargs, Expr(:kw, :($r::Union{arb_rnd,RoundingMode}), :(RoundNearest)))
-        elseif c_types[k] == Base.MPFR.MPFRRoundingMode
+        # Automatic detection of precision argument
+        if carg == Carg{Clong}(:prec, false)
+            @assert !prec_kwarg
+            prec_kwarg = true
+
+            c₁ = first(cargs)
+            push!(kwargs, Expr(:kw, :(prec::Integer), :(_precision($(name(c₁))))))
+
+            # Automatic detection of rounding mode argument
+        elseif carg == Carg{arb_rnd}(:rnd, false)
+            @assert !rnd_kwarg
+            rnd_kwarg = true
+
+            push!(
+                kwargs,
+                Expr(:kw, :(rnd::Union{$(arb_rnd),RoundingMode}), :(RoundNearest)),
+            )
+        elseif carg == Carg{Base.MPFR.MPFRRoundingMode}(:rnd, false)
+            @assert !rnd_kwarg
+            rnd_kwarg = true
             push!(
                 kwargs,
                 Expr(
                     :kw,
-                    :($r::Union{Base.MPFR.MPFRRoundingMode,RoundingMode}),
+                    :(rnd::Union{Base.MPFR.MPFRRoundingMode,RoundingMode}),
                     :(RoundNearest),
                 ),
             )
+            # Automatic detection of length arguments for vectors
+        elseif (startswith(string(name(carg)), "len") || carg == Carg{Clong}(:n, false)) &&
+               rawtype(cargs[i-1]) ∈ (ArbVector, AcbVector) &&
+               name(carg) ∉ len_keywords
+
+            vec_name = name(cargs[i-1])
+            push!(kwargs, Expr(:kw, :($(name(carg))::Integer), :(length($vec_name))))
+            push!(len_keywords, name(carg))
+
+        else
```

Comment:
Could we factor this out as a separate method? Mainly to make it easier to test independently since this is quite delicate.

Reply:
I factored it out and added some tests